cmd/xl-storage.go

```diff
@@ -961,17 +961,33 @@ func (s *xlStorage) DeleteVersions(ctx context.Context, volume string, versions
 	return errs
 }
 
-func (s *xlStorage) moveToTrash(filePath string, recursive bool) error {
+func (s *xlStorage) moveToTrash(filePath string, recursive, purge bool) error {
 	pathUUID := mustGetUUID()
+	targetPath := pathutil.Join(s.diskPath, minioMetaTmpDeletedBucket, pathUUID)
+
+	var renameFn func(source, target string) error
 	if recursive {
-		return renameAll(filePath, pathutil.Join(s.diskPath, minioMetaTmpDeletedBucket, pathUUID))
+		renameFn = renameAll
+	} else {
+		renameFn = Rename
+	}
+
+	if err := renameFn(filePath, targetPath); err != nil {
+		return err
 	}
-	return Rename(filePath, pathutil.Join(s.diskPath, minioMetaTmpDeletedBucket, pathUUID))
+
+	if purge {
+		go func() {
+			removeAll(targetPath)
+		}()
```

Comment:
need to benchmark this between parallel and sequential `os.RemoveAll` calls

Reply:
We should never do this @Praveenrajmani it should be blocking call.. otherwise this can lead to uncontrolled calls on the drive leading to lot more IOPs loss.